internal/service/service.go

```diff
@@ -0,0 +1,159 @@
+// Copyright  observIQ, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/observiq/observiq-otel-collector/collector"
+	"go.uber.org/zap"
+)
+
+const (
+	startTimeout = 10 * time.Second
+	stopTimeout  = 10 * time.Second
+)
+
+// RunnableService may be run as a service.
+type RunnableService interface {
+	// Start asynchronously starts the underlying service. The service may not necessarily be "ready"
+	// once this returns, but could be asynchronously starting up.
+	Start(ctx context.Context) error
+	// Stop synchronously shuts down the service. After this function returns, the underlying service should be completely stopped.
+	Stop(ctx context.Context) error
+	// Error returns an error channel that should emit an error when the service must unexpectedly quit.
+	Error() <-chan error
+}
+
+// StandaloneCollectorService is a RunnableService that runs the collector in standalone mode.
+type StandaloneCollectorService struct {
+	col      collector.Collector
+	doneChan chan struct{}
+	errChan  chan error
+	wg       *sync.WaitGroup
+}
+
+// NewStandaloneCollectorService creates a new StandaloneCollectorService
+func NewStandaloneCollectorService(c collector.Collector) StandaloneCollectorService {
+	return StandaloneCollectorService{
+		col:      c,
+		doneChan: make(chan struct{}, 1),
+		errChan:  make(chan error, 1),
+		wg:       &sync.WaitGroup{},
+	}
+}
+
+// Start starts the collector
+func (s StandaloneCollectorService) Start(ctx context.Context) error {
+	collectorStartedChan := make(chan struct{})
+	go func() {
+		// Start's context is only valid for the lifetime of "Start",
+		// but the collector expects a context that is valid for the lifetime of the service.
+		err := s.col.Run(context.Background())
```

Comment:
The comment on the context here confuses me. Why is `Start`'s context only good for the lifetime of start? How do we pass in a context that is listening for signals?

Reply:
The context for start contains timeout information (the Start operation should be quick and not long-lasting, I want to bail out and restart if it hangs).

Maybe it should be long running, though. I think the timeout could be handled outside this function.

Either way, Shutdown WILL be called when a signal is caught, which I think we need to do anyways to ensure the collector is fully shut down before exiting the program.

Reply:
I saw this later on but I still think we can use the signal context as a parent. See https://github.com/observIQ/observiq-otel-collector/pull/433#discussion_r878097759

Reply:
I kept it with the short running ctx (signal context becomes the parent). It becomes overcomplicated and doesn't really seem necessary, since Shutdown is called. Let me know if there's strong disagreement with that.

Reply:
Refactored to be long-running after discussing on a call.